src/animation.rs

```diff
@@ -0,0 +1,503 @@
+//! # let mut window = three::Window::builder("", "").build();
+//! # let camera = unimplemented!();
+//! # let mut mixer = three::animation::Mixer::new();
+//! # let (group, _, _, clips) = window.factory.load_gltf("AnimatedScene.gltf");
+//! # window.scene.add(&group);
+//! # let actions: Vec<three::animation::Action> = clips
+//! #     .into_iter()
+//! #     .map(|clip| mixer.action(clip))
+//! #    .collect();
+//! while window.update() {
+//!     mixer.update(window.input.delta_time());
+//!     window.render(&camera);
+//! }
+//! ```
+//!
+//! ### Putting it all together
+//!
+//! See the `gltf-animation` example for the full code.
+//!
+//! [`disable`]: struct.Action.html#method.disable
+//! [`play`]: struct.Action.html#method.play
+//! [`pause`]: struct.Action.html#method.pause
+//!
+//! [`Action`]: struct.Action.html
+//! [`Clip`]: struct.Clip.html
+//! [`Mixer`]: struct.Mixer.html
+//! [`Mixer::action`]: struct.Mixer.html#method.action
+//! [`Mixer::update`]: struct.Mixer.html#method.update
+
+use cgmath;
+use froggy;
+use mint;
+use std::sync::mpsc;
+
+use Object;
+
+/// Describes the interpolation behaviour between keyframes.
+#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
+pub enum Interpolation {
+    /// Immediate change between keyframe values.
+    Discrete,
+
+    /// Linear interpolation between keyframe values.
+    Linear,
+
+    /// Smooth cubic interpolation between keyframe values.
+    Cubic,
+
+    /// Smooth Catmull–Rom spline interpolation between keyframe values.
+    CatmullRom,
+}
+
+/// Describes the looping behaviour of an [`Action`].
+///
+/// [`Action`]: struct.Action.html
+#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
+pub enum LoopMode {
+    /// Play the clip exactly once, i.e. do not loop at all.
+    Once,
+
+    /// Play the clip with the chosen number of repetitions.
+    Repeat(Repetitions),
+
+    /// Play the clip with the chosen number of repetitions, alternatively
+    /// playing in forward and reverse order.
+    PingPong(Repetitions),
+}
+
+/// Describes the number of times an [`Action`] repeats itself.
+///
+/// [`Action`]: struct.Action.html
+#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
+pub enum Repetitions {
+    /// Repeat an exact number of times.
+    Finite(i32),
+
+    /// Repeat indefinately.
+    Infinite,
+}
+
+/// Describes the target property of an animation.
+#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
+pub enum Binding {
+    /// Targets the position property of an [`Object`].
+    ///
+    /// The corresponding keyframe values must be [`Vector3`].
+    ///
+    /// [`Object`]: ../object/struct.Object.html
+    /// [`Vector3`]: enum.Values.html#variant.Vector3
+    Position,
+
+    /// Targets the orientation property of an [`Object`].
+    ///
+    /// The corresponding keyframe values must be [`Quaternion`].
+    ///
+    /// [`Object`]: ../object/struct.Object.html
+    /// [`Quaternion`]: enum.Values.html#variant.Quaternion
+    Orientation,
```

Comment:
should we differentiate between quaternion binding and euler binding?

Reply:
That's a good point. I'm tempted to not support Euler angles due to the possibility of gimbal lock. Do you think it's an important feature to have?

Reply:
Yes, some 3D suites / animator workflows would produce euler angle clips. I don't think gimbal lock is a concern here - the 3D suite can make sure that the produced clips are valid, it's not our task to validate.